File: backend/app/services/command_center_service.py
"""
FindSafe AI - Command Center Intelligence Service (Phase 14)
Aggregates live platform statistics, case status distributions, evidence breakdown,
case spotlight, camera coverage GIS metadata, and real-time activity streams.
"""

import logging
from typing import List, Dict, Any, Optional
from datetime import datetime, timezone

logger = logging.getLogger("findsafe.ai.command_center_service")


class CommandCenterService:
    """Service providing high-level operational intelligence aggregations for Authority Command Center."""

    @staticmethod
    def get_dashboard_summary(supabase: Any) -> Dict[str, Any]:
        """
        Queries real database state across missing_persons, search_sessions, candidate_groups,
        found_person_records, and reports to compute live Command Center metrics.
        """
        if not supabase:
            return CommandCenterService._get_empty_summary()

        try:
            # 1. Live Missing Person Cases
            cases_res = supabase.table("missing_persons") \
                .select("id, case_id, full_name, case_title, status, priority, reference_image_url, reference_image_path, last_seen_location, last_seen_date, age, gender, created_at, updated_at") \
                .order("updated_at", desc=True) \
                .execute()
            all_cases = cases_res.data or []

            active_cases = len([c for c in all_cases if (c.get("status") or "").lower() not in ["closed", "archived", "resolved"]])

            status_dist = {"active": 0, "under_review": 0, "resolved": 0, "closed": 0, "archived": 0}
            priority_dist = {"low": 0, "medium": 0, "high": 0, "critical": 0}

            for c in all_cases:
                st = (c.get("status") or "active").lower()
                pr = (c.get("priority") or "medium").lower()
                if st in status_dist:
                    status_dist[st] += 1
                else:
                    status_dist["active"] += 1

                if pr in priority_dist:
                    priority_dist[pr] += 1
                else:
                    priority_dist["medium"] += 1

            # 2. Live Search Sessions
            searches_res = supabase.table("search_sessions") \
                .select("id, status, search_type, case_id, created_at, updated_at") \
                .order("created_at", desc=True) \
                .execute()
            all_searches = searches_res.data or []

            active_searches = len([s for s in all_searches if (s.get("status") or "").lower() in ["pending", "uploading", "processing"]])
            partial_searches = len([s for s in all_searches if (s.get("status") or "").lower() == "partial"])

            search_dist = {"processing": 0, "completed": 0, "partial": 0, "failed": 0, "cancelled": 0}
            for s in all_searches:
                st = (s.get("status") or "completed").lower()
                if st in search_dist:
                    search_dist[st] += 1
                elif st in ["pending", "uploading"]:
                    search_dist["processing"] += 1

            # 3. Live Candidate Groups & Reviews
            groups_res = supabase.table("candidate_groups") \
                .select("id, status, overall_score, visual_score, evidence_level, case_id, created_at") \
                .order("created_at", desc=True) \
                .execute()
            all_groups = groups_res.data or []
            candidate_groups_count = len(all_groups)
            pending_reviews = len([g for g in all_groups if (g.get("status") or "").lower() in ["under_review", "potential_match"]])

            # 4. Live Records & Reports
            records_count = 0
            try:
                rec_res = supabase.table("found_person_records").select("id", count="exact").execute()
                records_count = len(rec_res.data or [])
            except Exception:
                pass

            reports_count = 0
            try:
                rep_res = supabase.table("investigation_reports").select("id", count="exact").execute()
                reports_count = len(rep_res.data or [])
            except Exception:
                pass

            # 5. Case Spotlight (Most Recently Active Missing Person Case)
            spotlight_case = None
            if all_cases:
                top_case = dict(all_cases[0])
                # Format reference photo URL
                ref_path = top_case.get("reference_image_path")
                ref_url = top_case.get("reference_image_url")
                if not ref_url and ref_path:
                    try:
                        pub = supabase.storage.from_("missing-person-photos").get_public_url(ref_path)
                        top_case["reference_image_url"] = pub.get("publicUrl") if isinstance(pub, dict) else str(pub)
                    except Exception:
                        top_case["reference_image_url"] = ref_path

                spotlight_case = {
                    "id": top_case.get("id"),
                    "case_id": top_case.get("case_id") or top_case.get("id"),
                    "case_title": top_case.get("case_title") or top_case.get("full_name") or "Missing Person Case",
                    "full_name": top_case.get("full_name") or top_case.get("case_title") or "Unknown Subject",
                    "status": top_case.get("status") or "active",
                    "priority": top_case.get("priority") or "high",
                    "age": top_case.get("age"),
                    "gender": top_case.get("gender"),
                    "last_seen_location": top_case.get("last_seen_location") or "Reported Missing",
                    "last_seen_date": top_case.get("last_seen_date") or top_case.get("created_at"),
                    "reference_image_url": top_case.get("reference_image_url"),
                    "candidate_count": candidate_groups_count,
                    "updated_at": top_case.get("updated_at") or top_case.get("created_at")
                }

            # 6. Live Review Queue
            review_queue = []
            case_map = {c.get("id"): c for c in all_cases}
            for grp in all_groups[:6]:
                cid = grp.get("case_id")
                c_item = case_map.get(cid, {})
                review_queue.append({
                    "id": grp.get("id"),
                    "case_id": cid,
                    "case_title": c_item.get("case_title") or c_item.get("full_name") or "Investigation Review",
                    "case_priority": c_item.get("priority") or "high",
                    "evidence_score": float(grp.get("overall_score") or 0.0),
                    "evidence_level": grp.get("evidence_level") or "moderate",
                    "status": grp.get("status") or "potential_match",
                    "created_at": grp.get("created_at") or datetime.now(timezone.utc).isoformat()
                })

            # 7. Live Activity Feed from real events
            activity_feed = []
            for s in all_searches[:5]:
                cid = s.get("case_id")
                c_item = case_map.get(cid, {})
                activity_feed.append({
                    "id": f"act-search-{s.get('id')}",
                    "action": f"{s.get('search_type', 'Crowd').capitalize()} Search {s.get('status', 'Completed').capitalize()}",
                    "performed_by": "System AI Engine",
                    "timestamp": s.get("updated_at") or s.get("created_at"),
                    "details": {
                        "label": f"Session for {c_item.get('full_name') or 'Case File'}",
                        "status": s.get("status")
                    }
                })

            for c in all_cases[:3]:
                activity_feed.append({
                    "id": f"act-case-{c.get('id')}",
                    "action": "Case Registered / Updated",
                    "performed_by": "Authority Operator",
                    "timestamp": c.get("updated_at") or c.get("created_at"),
                    "details": {
                        "label": f"{c.get('full_name') or c.get('case_title')} ({c.get('priority', 'Normal')})",
                        "status": c.get("status")
                    }
                })

            activity_feed.sort(key=lambda x: str(x.get("timestamp") or ""), reverse=True)

            return {
                "summary_strip": {
                    "active_cases": active_cases,
                    "active_searches": active_searches,
                    "pending_reviews": pending_reviews,
                    "open_tasks": pending_reviews + active_searches,
                    "partial_searches": partial_searches,
                    "reports_count": reports_count,
                    "system_status": "Operational"
                },
                "case_status_distribution": status_dist,
                "case_priority_distribution": priority_dist,
                "search_status_distribution": search_dist,
                "evidence_overview": {
                    "candidate_groups": candidate_groups_count,
                    "record_matches": records_count,
                    "cross_source_associations": min(candidate_groups_count, records_count) + (1 if candidate_groups_count > 0 and records_count > 0 else 0),
                    "reports_generated": reports_count
                },
                "evidence_sources": {
                    "camera": candidate_groups_count,
                    "records": records_count,
                    "reports": reports_count
                },
                "spotlight_case": spotlight_case,
                "cases": all_cases[:10],
                "review_queue": review_queue,
                "activity_feed": activity_feed[:10]
            }

        except Exception as e:
            logger.error(f"Error compiling live command center summary: {str(e)}", exc_info=True)
            return CommandCenterService._get_empty_summary()

    @staticmethod
    def get_camera_coverage(supabase: Any) -> List[Dict[str, Any]]:
        """
        Retrieves camera sources used across search sessions with location coordinates,
        usage timestamps, and operational status labels.
        """
        if not supabase:
            return []

        try:
            # Query camera sources and search session videos
            res = supabase.table("search_session_videos").select("id, camera_name, latitude, longitude, created_at, processing_status").order("created_at", desc=True).execute()
            svideos = res.data or []

            cameras = []
            seen_cams = set()

            for idx, sv in enumerate(svideos):
                cam_name = sv.get("camera_name") or f"Camera {idx+1:02d}"
                if cam_name in seen_cams:
                    continue
                seen_cams.add(cam_name)

                lat = sv.get("latitude") or (28.6139 + (idx * 0.003))
                lng = sv.get("longitude") or (77.2090 + (idx * 0.003))

                cameras.append({
                    "id": sv.get("id"),
                    "camera_name": cam_name,
                    "location_name": f"Surveillance Sector {idx + 1}",
                    "lat": float(lat),
                    "lng": float(lng),
                    "status_label": "Recently Used" if idx == 0 else "Available",
                    "last_used": sv.get("created_at", "Recently"),
                    "total_searches": idx + 1
                })

            return cameras
        except Exception as e:
            logger.error(f"Error fetching camera coverage: {str(e)}")
            return []

    @staticmethod
    def _get_empty_summary() -> Dict[str, Any]:
        """Provides default empty structure when database is empty."""
        return {
            "summary_strip": {
                "active_cases": 0,
                "active_searches": 0,
                "pending_reviews": 0,
                "open_tasks": 0,
                "partial_searches": 0,
                "reports_count": 0,
                "system_status": "Operational"
            },
            "case_status_distribution": {"active": 0, "under_review": 0, "resolved": 0, "closed": 0, "archived": 0},
            "case_priority_distribution": {"low": 0, "medium": 0, "high": 0, "critical": 0},
            "search_status_distribution": {"processing": 0, "completed": 0, "partial": 0, "failed": 0, "cancelled": 0},
            "evidence_overview": {
                "candidate_groups": 0,
                "record_matches": 0,
                "cross_source_associations": 0,
                "reports_generated": 0
            },
            "evidence_sources": {"camera": 0, "records": 0, "reports": 0},
            "spotlight_case": None,
            "cases": [],
            "review_queue": [],
            "activity_feed": []
        }

File: frontend/src/services/commandCenterService.ts
import { apiClient } from './api';

export interface SummaryStrip {
  active_cases: number;
  active_searches: number;
  pending_reviews: number;
  open_tasks: number;
  partial_searches: number;
  reports_count: number;
  system_status: string;
}

export interface CommandCenterSummaryData {
  summary_metrics?: {
    active_cases: number;
    active_searches: number;
    pending_reviews: number;
    open_tasks: number;
    partial_searches: number;
    reports_count: number;
    system_status: string;
  };
  summary_strip?: SummaryStrip;
  case_status_distribution?: { [key: string]: number };
  case_priority_distribution?: { [key: string]: number };
  search_status_distribution?: { [key: string]: number };
  evidence_overview?: {
    camera_evidence: number;
    record_evidence: number;
    cross_source_associations: number;
    candidate_groups: number;
    reports: number;
  };
  source_distribution?: { [key: string]: number };
  case_spotlight?: any;
  cases?: any[];
  review_queue?: any[];
  investigator_workload?: any[];
  task_stats?: {
    open: number;
    completed: number;
    overdue: number;
  };
  recent_activity?: any[];
}

export type CommandCenterSummary = CommandCenterSummaryData;

export interface CameraCoverageItem {
  id: string;
  camera_name?: string;
  location_name?: string;
  name?: string;
  location?: string;
  lat: number;
  lng: number;
  status_label?: string;
  status?: string;
  last_used?: string;
  total_searches?: number;
  searches_count?: number;
}

export const commandCenterService = {
  getSummary: async (): Promise<CommandCenterSummaryData> => {
    const res = await apiClient.get('/command-center/summary');
    const data = res.data;
    // Map backend JSON fields to frontend convenient structure if needed
    return {
      summary_metrics: data.summary_strip || {
        active_cases: data.summary_strip?.active_cases ?? 0,
        active_searches: data.summary_strip?.active_searches ?? 0,
        pending_reviews: data.summary_strip?.pending_reviews ?? 0,
        open_tasks: data.summary_strip?.open_tasks ?? 0,
        partial_searches: data.summary_strip?.partial_searches ?? 0,
        reports_count: data.summary_strip?.reports_count ?? 0,
        system_status: data.summary_strip?.system_status || 'Operational'
      },
      summary_strip: data.summary_strip,
      case_status_distribution: data.case_status_distribution || {},
      case_priority_distribution: data.case_priority_distribution || {},
      search_status_distribution: data.search_status_distribution || {},
      evidence_overview: {
        camera_evidence: data.evidence_overview?.candidate_groups ?? 0,
        record_evidence: data.evidence_overview?.record_matches ?? 0,
        cross_source_associations: data.evidence_overview?.cross_source_associations ?? 0,
        candidate_groups: data.evidence_overview?.candidate_groups ?? 0,
        reports: data.evidence_overview?.reports_generated ?? 0,
      },
      source_distribution: data.evidence_sources || {},
      case_spotlight: data.spotlight_case || null,
      cases: data.cases || [],
      review_queue: data.review_queue || [],
      investigator_workload: data.investigator_workload || [
        { investigator: 'Lead Investigator', role: 'SUPERVISOR', assigned_cases: data.summary_strip?.active_cases || 1, pending_reviews: data.summary_strip?.pending_reviews || 0, open_tasks: data.summary_strip?.open_tasks || 0 }
      ],
      task_stats: data.task_stats || { open: data.summary_strip?.open_tasks || 0, completed: 0, overdue: 0 },
      recent_activity: data.activity_feed || []
    };
  },

  getCameras: async (): Promise<CameraCoverageItem[]> => {
    const res = await apiClient.get('/command-center/cameras');
    return (res.data || []).map((c: any) => ({
      id: c.id,
      name: c.camera_name || c.name || 'Camera Feed',
      location: c.location_name || c.location || 'Location Ground',
      lat: c.lat || 12.9716,
      lng: c.lng || 77.5946,
      status: c.status_label || c.status || 'Available',
      searches_count: c.total_searches ?? c.searches_count ?? 1,
      last_used: c.last_used || new Date().toISOString()
    }));
  }
};

